provider/core/src/resource.rs

```diff
@@ -79,10 +80,62 @@ impl writeable::Writeable for ResourceCategory {
 ///
 /// Use [`resource_key!`] as a shortcut to create resource keys in code.
 #[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+#[repr(C)]
 pub struct ResourceKey {
+    _tag0: TinyStr8,
     pub category: ResourceCategory,
     pub sub_category: TinyStr16,
-    pub version: u16,
+    pub version: u8,
+    _tag1: TinyStr4,
+}
+
+impl ResourceKey {
+    /// Creates a new [`ResourceKey`].
+    pub const fn new(
+        category: ResourceCategory,
+        sub_category: TinyStr16,
+        version: u8,
+    ) -> ResourceKey {
+        ResourceKey {
+            _tag0: tinystr8!("ICURES[["),
```

Comment:
Is this going to work for a BE binary?

Reply:
It will work so long as the fields are endian-agnostic.  I changed the version field to be `u8` instead of `u16` to address this problem.  The only field I'm not sure about is `ResourceCategory`, but I left a TODO to fix that.

Reply:
For TinyStr in particular, there's a conversation about it.  Short answer is that it actually is endian-safe.

https://github.com/zbraniecki/tinystr/issues/45

Reply:
For ResourceCategory it will be fine as long as it's `repr(u8)`.

Reply:
ResourceCategory is dataful, which is why it is hard.  The other problem is that I intend to remove ResourceCategory, so I don't want to spend a lot of time solving it.  Also, even if ResourceCategory were `repr(u8)`, it's still not safe to transmute, because the u8 could be out-of-range.

What do you suggest as the short-term solution?

1. Check in this unsafe code first and remove ResourceCategory later
2. Remove ResourceCategory first before merging this PR
3. Make ResourceCategory safe, and then delete that code when I remove ResourceCategory

Reply:
Ah, I did not realize ResourceCategory is dataful.

I'm okay with having this code for now but I'd prioritize fixing this up.

@robertbastian yes, because it has the same bit pattern as the string, and strings are endianness-independent. In fact I've thought about making tinystr internally use byte arrays in the first place.

Reply:
I still think it overcomplicates things to write a `TinyStr` but then compare to a byte array.